devito/types/basic.py

```diff
@@ -423,8 +434,113 @@ def __dtype_setup__(cls, **kwargs):
         return kwargs.get('dtype', np.float32)
 
 
-class AbstractFunction(sympy.Function, Basic, Cached, Pickable, Evaluable):
+class AbstractTensor(sympy.ImmutableDenseMatrix, Basic, Cached, Pickable, Evaluable):
+    """
+    Base class for vector and tensor valued functions. It inherits from and
+    mimicks the behavior of a sympy.ImmutableDenseMatrix.
+
+
+    The sub-hierachy is as follows
+
+                         AbstractTensor
+                                |
+                      AbstractCachedTensor
```

Comment:
Look at the new AbstractFunction. There is no AbstractCachedFunction any more. Same here -- we can drop AbstractCachedTensor and make TensorFunction inherit directly from AbstractTensor

EDIT : I think you've dropped it already. You just have to clean up the docstring. Please run a `git grep AbstractCachedTensor` to make sure all occurrences have been deleted